### ISSUE TITLE
receive_exactly causing infinite loop if stream is closed?

### ISSUE THREAD

**Issue by mzealey** [agronholm/anyio#162]
https://github.com/agronholm/anyio/blob/3e3f988abdd97435d06e57fe183a53d230ab0142/src/anyio/streams/buffered.py#L71

I'm using a TCP TLS wrapped stream and receiving packets on it. However if the stream was closed unexpectedly with 0 bytes left in the buffer it appears that we can get stuck in an infinite loop here - perhaps a read() returning 0 bytes should flag the connection as closed or something?

**Comment by agronholm** [comment]
If the stream has been closed, reading from it raises `anyio.EndOfStream`. Are you however getting an infinite loop in practice?

**Comment by mzealey** [comment]
Yes I am getting an infinite loop from calling receive_exactly if the tcp/tls stream is closed under some situations (not too sure exactly but probably a timing/race thing). if its a byte stream returning a 0 read i thought that was an EOF signal usually?

**Comment by agronholm** [comment]
All of AnyIO's TCP streams raise that exception when they read 0 bytes, or they at least should. If not, that's a bug. Can you show me how to reproduce this locally?

**Comment by agronholm** [comment]
Also, which backend were you using? That might give me some additional clues on where to look for problems.

**Comment by mzealey** [comment]
See https://github.com/mzealey/anyio-hang

Get a tls server cert/key (anything will do)

```
docker build -t dot-test . && docker run -it --rm --init -it -v $PWD/../test-certs:/cert -e CERT_FILE=/cert/server.crt -e KEY_FILE=/cert/server.key dot-test
```

using kdig from knot-utils:

```
kdig +short @172.17.0.3 -p 853 +tls +tcp foo.bar
```

Correctly returns CNAME foo.bar. as response but then hangs in infinite loop at the point marked.

**Comment by agronholm** [comment]
Ah, so the likely culprit is the `TLSStream` class, not TCP streams. I will try to repro and investigate then.

**Comment by agronholm** [comment]
I can see 100% CPU utilization after the response. TLSStream is definitely where the problem occurs though. I will need to add test cases to cover this once I understand why this happens.

**Comment by mzealey** [comment]
Many thanks, confirm this fixes my issue. I now receive `anyio.IncompleteRead` on ssl connection termination when doing receive_exactly

**Comment by agronholm** [comment]
I have a third bug fix for v2.0 coming shortly, and after that I will release v2.0.1.
